llvm/lib/Transforms/Scev/Scev.cpp

```diff
@@ -0,0 +1,110 @@
+/*===- Scev.cpp -Creates and Simplifies Recurrences for ‘Expressions involving
+Induction Variables’ Algorithm:
+1. Get ScalarEvolution object.
+2. Use getSCEV for the pointer operands
+3. Take the scev pointer base
+4. Subtract scev with scev pointer base to get the SCEVAddRecExpr(DiffVal).
+eg:{8,+,16}<nuw><nsw><%for.cond>
+5. This SCEVAddRecExpr will contain the required indices and Extract it. eg : 8
+6. Store the index and corresponding Store instruction in StoreInsts map.
+7. Sorting the Offset vector values.
+8. Get the BB of store instruction and using that get the terminator
+instruction.
+9. Move all store instructions one by one before terminator instruction.
+===-------------------------------------------------------------------------------------------===*/
+
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/Statistic.h"
+#include "llvm/Analysis/ScalarEvolution.h"
+#include "llvm/Analysis/ScalarEvolutionExpressions.h"
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/CFG.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/Instruction.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/raw_ostream.h"
+#include <algorithm>
+using namespace llvm;
+
+#define DEBUG_TYPE "hello"
+
+namespace {
+// Scev - The second implementation with getAnalysisUsage implemented.
+struct Scev : public FunctionPass {
+  static char ID; // Pass identification, replacement for typeid
+  Scev() : FunctionPass(ID) {}
+
+  bool runOnFunction(Function &F) override {
+
+    auto &SE = getAnalysis<ScalarEvolutionWrapperPass>().getSE();
+    SCEV *ScevVal, *BasePtr, *DiffVal, *GetEle, *TempPtr = nullptr;
+    SmallVector<int> OffSet;
+    llvm::DenseMap<int, Instruction *> StoreInsts;
+
+    int Value = 0;
+
+    // Store the index and corresponding Store instruction in StoreInsts map
+    for (BasicBlock &BB : F) {
+      for (Instruction &I : BB) {
+        if (auto *Store = dyn_cast<StoreInst>(&I)) {
+          if (auto *Gep =
+                  dyn_cast<GetElementPtrInst>(Store->getPointerOperand())) {
+            ScevVal = const_cast<SCEV *>(SE.getSCEV(Gep));
+            if ((BasePtr = const_cast<SCEV *>(SE.getPointerBase(ScevVal)))) {
+              if (TempPtr == nullptr)
```

Comment:
this will not work for all cases.
instead of checking prev and curr base pointer,use a mapVector<SCEV *,SmallVector<Instns>> 

at the end it will have vector of instructions corresponding to a base pointer.